# 2주차 강사용 Notebook — 파이썬과 데이터 분석의 기본

학생용 Notebook과 동일한 흐름이며, 예상 결과와 설명 포인트를 함께 담았습니다.

## 1단계. 온도 변수 만들기
**설명 포인트**: 변수는 값 하나만 담을 수 있다는 점을 강조. `type()`으로 자료형(float)을 확인.

In [1]:
temp1 = 302.7
print(temp1)
print(type(temp1))

302.7
<class 'float'>


## 2단계. 리스트에 여러 온도 저장하기
**설명 포인트**: 변수 하나로는 값을 여러 개 담을 수 없으므로 리스트가 필요하다는 점을 연결해서 설명.

In [2]:
temps = [302.7, 300.1, 292.0, 299.0, 291.9, 303.5]
print(temps)
print("측정 개수:", len(temps))

[302.7, 300.1, 292.0, 299.0, 291.9, 303.5]
측정 개수: 6


## 3단계. 평균/최댓값/최솟값 계산하기
**어려워할 부분**: `sum(temps) / len(temps)`에서 괄호 안에 리스트 이름을 넣어야 한다는 것을 놓치기 쉬움.

In [3]:
average = sum(temps) / len(temps)
maximum = max(temps)
minimum = min(temps)

print("평균:", round(average, 1), "최댓값:", maximum, "최솟값:", minimum)

평균: 298.2 최댓값: 303.5 최솟값: 291.9


**예상 결과**: 평균 298.2, 최댓값 303.5, 최솟값 291.9 (앞 6개 값 기준 — 전체 25건과 다름을 5단계에서 비교).

## 4단계. 정상범위(295~305℃) 조건 확인하기
**질문할 내용**: "or 대신 and를 쓰면 결과가 어떻게 달라질까요?" (극단적인 경우를 놓칠 수 있음을 설명)

In [4]:
if maximum > 305 or minimum < 295:
    print("정상범위를 벗어난 값이 있습니다.")
else:
    print("모든 값이 정상범위 안에 있습니다.")

정상범위를 벗어난 값이 있습니다.


## 5단계. 전체 데이터를 Pandas로 불러와 리스트로 바꾸기
**설명 포인트**: `df["온도_섭씨"].tolist()`는 Pandas 열(Series)을 우리가 배운 파이썬 리스트로 바꿔준다.
즉 1~4단계에서 손으로 만든 리스트를, Pandas가 CSV 파일 전체에 대해 대신 만들어주는 것이라는 연결고리를 강조.

In [5]:
import pandas as pd

df = pd.read_csv("../../data/weekly/week02/week02_basic_process_values.csv")
df.head(8)

,측정시간,로트번호,공정명,온도_섭씨,압력_Pa,합격여부
0,2024-03-01 10:00,LOT-0139,증착,302.7,1005.9,1
1,2024-03-01 16:00,LOT-0088,식각,300.1,1001.3,1
2,2024-03-01 21:00,LOT-0045,증착,292.0,998.9,1
3,2024-03-02 04:00,LOT-0170,포토,299.0,1021.4,1
4,2024-03-02 07:00,LOT-0138,식각,291.9,1011.8,1
5,2024-03-02 19:00,LOT-0122,세정,303.5,1015.5,1
6,2024-03-02 23:00,LOT-0013,세정,307.6,1027.7,1
7,2024-03-03 04:00,LOT-0013,증착,306.0,1017.9,-1


In [6]:
all_temps = df["온도_섭씨"].tolist()

print("전체 평균:", round(sum(all_temps) / len(all_temps), 1))
print("전체 최댓값:", max(all_temps), " 전체 최솟값:", min(all_temps))

전체 평균: 301.9
전체 최댓값: 320.0  전체 최솟값: 289.3


**예상 결과**: 전체 평균 301.9, 최댓값 320.0, 최솟값 289.3.

**강조할 점**: 최댓값 320.0℃가 나온 행을 찾아 합격여부를 확인시킨다 — 극단값이라도 합격 처리된 사례가 있다는 것을
직접 눈으로 보여주면 8주차(원인 후보) 학습의 밑거름이 된다.

In [7]:
# 최댓값이 나온 행 확인 (강사 시연용)
df[df["온도_섭씨"] == df["온도_섭씨"].max()]

,측정시간,로트번호,공정명,온도_섭씨,압력_Pa,합격여부
24,2024-03-08 23:00,LOT-0047,식각,320.0,1011.9,1


## 6단계. 수율(합격 비율) 계산하기
**어려워할 부분**: 마지막에 100을 곱하는 것을 잊어버리는 경우가 많음.

In [8]:
pass_count = len(df[df["합격여부"] == 1])
total_count = len(df)
yield_pct = pass_count / total_count * 100

print(f"수율: {pass_count}/{total_count} = {round(yield_pct, 1)}%")

수율: 23/25 = 92.0%


**예상 결과**: 23/25 = 92.0%.

## 7단계(도전). 압력 리스트와 합격/불합격 그룹 비교하기
**실습 시간 조절**: 시간이 부족하면 이 단계는 숙제로 돌려도 된다.

In [9]:
pressures = df["압력_Pa"].tolist()
print("압력 평균:", round(sum(pressures) / len(pressures), 1))
print("압력 최댓값:", max(pressures), " 압력 최솟값:", min(pressures))

압력 평균: 1010.6
압력 최댓값: 1027.7  압력 최솟값: 992.3


In [10]:
pass_avg_temp = df[df["합격여부"] == 1]["온도_섭씨"].mean()
fail_avg_temp = df[df["합격여부"] == -1]["온도_섭씨"].mean()

print("합격 그룹 평균 온도:", round(pass_avg_temp, 1))
print("불합격 그룹 평균 온도:", round(fail_avg_temp, 1))

합격 그룹 평균 온도: 302.2
불합격 그룹 평균 온도: 297.6


**예상 결과**: 압력 평균 1010.6 / 최댓값 1027.7 / 최솟값 992.3.
합격 그룹 평균 온도 302.2℃, 불합격 그룹 평균 온도 297.6℃ — 불합격 그룹의 평균이 오히려 더 낮게 나오는데,
이는 불합격 2건 중 하나(289.3℃)가 매우 낮은 값이기 때문이다. 표본이 2건뿐이라 평균만으로 결론 내리기 어렵다는
점도 함께 짚어준다(6주차 groupby, 8주차 상관관계 예고).

## 8단계. 오류 대처 방법
- `NameError` 발생 시: 리스트/변수를 만드는 셀을 먼저 실행했는지 확인.
- `TypeError` 발생 시: 숫자와 문자열을 그냥 더하지 않았는지 확인 (f-문자열 사용 권장).
- `ZeroDivisionError` 발생 시: 리스트가 비어있지 않은지 `print()`로 먼저 확인.

## 확장 실습(빠른 학습자용)
`df.describe()`로 숫자 열 전체의 평균·표준편차·최댓값·최솟값을 한 번에 확인해보게 하면
3주차 이후 배우는 `groupby`, `describe()` 활용의 자연스러운 예고가 된다.

In [11]:
df.describe()

,온도_섭씨,압력_Pa,합격여부
count,25.000000,25.000000,25.000000
mean,301.872000,1010.576000,0.840000
std,7.875623,8.926659,0.553775
min,289.300000,992.300000,-1.000000
25%,297.900000,1005.900000,1.000000
50%,301.300000,1011.800000,1.000000
75%,303.500000,1016.200000,1.000000
max,320.000000,1027.700000,1.000000
